### ISSUE TITLE
Mario Party 3 (U): minigame Curtain Call issues

### ISSUE THREAD

**Issue by Mastergatto** [gonetz/GLideN64#692]
The character you select appears in the cloud, but the other choices are still visible outside of it, in line. 

![marioparty3u](https://cloud.githubusercontent.com/assets/1028191/9567671/40e8fca6-4f34-11e5-8da9-5c99ed060614.png)

Since I can't use the software rendering on Linux, so here is this preview of the minigame:
![curtaincall](https://cloud.githubusercontent.com/assets/1028191/9567683/870e8bce-4f34-11e5-9062-c5d8336227d5.png)

Here is the save state:
https://mega.nz/#!u4xFEDzR!7Z9xCdev1YGukH9t-3hsQpEM9KRpm5jwDhp3h7eexDA

**Comment by purplemarshmallow** [comment]
correct with angrylion's plugin
No other plugin seems to be correct

**Comment by olivieryuyu** [comment]
It works with N64 depth compares option

Fixed :)

**Comment by purplemarshmallow** [comment]
Yes it really works with N64 depth compare that's good
>Fixed

But can it be enabled in the ini? This option is very demanding. Performance is also important. Does the rest of the game work correctly with N64 depth compare?

**Comment by olivieryuyu** [comment]
it can be implemented in ini. However it will slow down for sure. The implementation of this depth compare is working per triangle as far as I understood. May be there is way to speed up the matter but Sergey got not luck here yet.

**Comment by purplemarshmallow** [comment]
LLE is also slow (compared with z64gl) because triangles are not optimized. If one optimization could solve both that would make GLideN64 clearly the best plugin available superior to everything else.

**Comment by purplemarshmallow** [comment]
I noticed in HLE N64 depth compare completely kills performance but in LLE it makes hardly any difference. I think with the right optimizations both LLE and N64 depth compare will be fast

**Comment by gonetz** [comment]
> If one optimization could solve both
It can't. LLE optimization is to draw as many triangles with one GL command as possible. That kills N64 depth compare shader.

**Comment by olivieryuyu** [comment]
please close as it is fixed

**Comment by purplemarshmallow** [comment]
it's not fixed with default settings
And I don't recommed setting it in the ini. I can't run it fullspeed with N64 depth compare enabled.

**Comment by purplemarshmallow** [comment]
>It can't. LLE optimization is to draw as many triangles with one GL command as possible. That kills N64 depth compare shader.

@gonetz can you elaborate this? I'm very interesten in N64 depth compare+optimized LLE. I can't see how it kills the depth compare shader but maybe I'm missing something?

**Comment by gonetz** [comment]
@purplemarshmallow N64 depth compare works well only when triangles have been sent one by one to GL processing. When draw command sends several polygons at once to draw, collisions occurred because synchronization of reads/writes to depth texture does not work between polygons. I tried all possible synchronization commands, nothing helped.

**Comment by standard-two-simplex** [comment]
@gonetz I had an idea to try to syncronize the load and writes and it turns out that it is being used quite commonly. It is based in a semaphore system: the semaphore is put to red light before read and to green after a write. While the semaphore is in red, no reads are allowed.

Here is a link to a stack overflow question with a similar situation. 
http://stackoverflow.com/questions/10652862/glsl-semaphores 
Did you try something like this?

Could you tell me how to send the triangles all together so I can test if it would work?

**Comment by gonetz** [comment]
It is easy: in gSPTriangle remove or comment out these two lines:
```
		if (config.frameBufferEmulation.N64DepthCompare != 0)
			render.drawTriangles();

```

**Comment by standard-two-simplex** [comment]
Meh. I implemented the semaphore but I can see no difference.

**Comment by gonetz** [comment]
The same result,  as I got :(

**Comment by standard-two-simplex** [comment]
I was looking around the internet when I stumbled upon [this](https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_fragment_shader_interlock.txt).

It looks like the extension needed to do synchronised writes to uniform images in an easy way. I'm pretty bad at OpenGL so I'm not sure how to add the extension nor if it requires updating GL version or anything of the like. Any ideas on how to use it?

**Comment by fzurita** [comment]
At the top of the shader add 
```
 #extension GL_ARB_fragment_shader_interlock : enable
```

Search the code for ```#extension``` for examples.

**Comment by gonetz** [comment]
I tried to use something like this to sync writes to depth image textures, without success.  May be someone can find a way to apply this technique properly.

**Comment by standard-two-simplex** [comment]
>At the top of the shader add
```
 #extension GL_ARB_fragment_shader_interlock : enable
```

@fzurita At the top of FragmentShaderHeader right? No conditions on OpenGL version are required?

> Could you tell me how to send the triangles all together so I can test if it would work?

How do I accomplish this in current code?

**Comment by gonetz** [comment]
For BufferedDrawer, in BufferedDrawer::drawTriangles
comment out
`if (config.frameBufferEmulation.N64DepthCompare == 0)`

For UnbufferedDrawer, in UnbufferedDrawer::drawTriangles
comment out 
`if (config.frameBufferEmulation.N64DepthCompare == 0)`

This will force drawing all polygons by one command. Desktop version usually uses BufferedDrawer.

**Comment by loganmc10** [comment]
> At the top of FragmentShaderHeader right? No conditions on OpenGL version are required?

You will need to check to make sure the graphics card supports that extension. I wouldn't worry about that for development, if you get it working I can take care of that part

**Comment by standard-two-simplex** [comment]
```
 #extension GL_ARB_fragment_shader_interlock : enable
```
If I write `` require`` instead of ``enable`` the shader doesn't compile. Does that mean my graphics card doesn't support it?

**Comment by fzurita** [comment]
Do a glxinfo command and see if you have ```ARB_fragment_shader_interlock``` listed.

**Comment by loganmc10** [comment]
You may also need:
```
#extension GL_ARB_shader_image_load_store : enable
#extension GL_ARB_shading_language_420pack : enable
```

The extension says:
> OpenGL 4.2 or ARB_shader_image_load_store is required; GLSL 4.20 is
    required.

GLideN64 requests an OpenGL 3.3 core profile, some graphics cards will return only 3.3, even if they support higher, some graphics cards will return the highest profile they support like 4.3. If it returns 3.3, then you need to enable those extensions to get access to ARB_shader_image_load_store  and GLSL 4.20

**Comment by standard-two-simplex** [comment]
> Do a glxinfo command and see if you have ARB_fragment_shader_interlock listed.

How do I do that?

```
#extension GL_ARB_shader_image_load_store : enable
#extension GL_ARB_shading_language_420pack : enable
```
Those 2 are already enabled in master, aren't they? I put them right before the other one anyway with no change.

My GPU supports OpenGL 4.5 so I'm surprised.

**Comment by loganmc10** [comment]
> Those 2 are already enabled in master, aren't they? I put them right before the other one anyway with no change.

Yeah they are there for the shaders that need them, I haven't been following along too closely so I don't know what shaders you're modifying

**Comment by fzurita** [comment]
The extension is not part of the OpenGL 4.5 standard, it's an extension to it. The glxinfo command is a command you type in a linux terminal window. I realize now that you are running Windows since I remember you were compiling using Visual Studio 2015.

I'm not sure of an easy way to see available extensions in Windows. I guess you could change the code to LOG all available extensions right here:
https://github.com/gonetz/GLideN64/blob/master/src/Graphics/OpenGLContext/opengl_Utils.cpp#L16

Also, you could do
```
Utils::isExtensionSupported("ARB_fragment_shader_interlock");
```
To see if you have the extension available.

**Comment by standard-two-simplex** [comment]
```
Utils::isExtensionSupported("ARB_fragment_shader_interlock");
```
That is returning false. So I guess I can't test the code.

**Comment by loganmc10** [comment]
It should have the GL_ prefix like ```GL_ARB_fragment_shader_interlock```

**Comment by standard-two-simplex** [comment]
Still false.

**Comment by Kidboy17** [comment]
The issue still happens on the newest WIP (Master+FB) released today with n64 depth compare enabled. Using Project 64 2.4.0.668-gc506db5.

**Comment by NikitaGamer64** [comment]
I can’t find the depth buffer to compare. How to?
